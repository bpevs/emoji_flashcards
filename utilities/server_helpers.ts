import { Context } from 'oak'
import { readLanguageFile } from './data_access.ts'
import { Handlebars } from 'handlebars'
import locales from '../data/locales.js'
import { DEFAULT_LANG, NOTE_PARAM, USER_PARAM } from './constants_shared.ts'

const handle = new Handlebars({
  cachePartials: true,
  compilerOptions: undefined,
  defaultLayout: 'main',
  extname: '.hbs',
  baseDir: 'www/views',
  helpers: {
    // deno-lint-ignore no-explicit-any
    select(selected: string, options: { fn: any }) {
      const findOptionValue = new RegExp(' value="' + selected + '"')
      // @ts-ignore need to figure out handlebar helper option types
      return options.fn(this).replace(findOptionValue, '$& selected="selected"')
    },
  },
  layoutsDir: 'layouts/',
  partialsDir: 'partials/',
})

export function pageHandler(pageName: string) {
  return async (ctx: Context) => {
    const data = await getWebsiteData(ctx.request.url.searchParams)
    ctx.response.body = await handle.renderView(pageName, data)
  }
}

export async function getWebsiteData(params: URLSearchParams): Promise<{
  categories: string[]
  userLangCode: string
  noteLangCode: string
  flag: string
  strings: { [name: string]: string }
  notes: string[][]
  locales: Array<{
    language_code: string
    locale_code: string
    native_name: string
    locale_flag: string
  }>
}> {
  const userLangCode = params.get(USER_PARAM) || DEFAULT_LANG
  const noteLangCode = params.get(NOTE_PARAM) || DEFAULT_LANG

  const { strings } = await readLanguageFile(userLangCode, false)
  const { data, locale_flag } = await readLanguageFile(noteLangCode, true)

  const categories = Object.keys(data).sort()
  const notes = categories.map((category: string) => {
    const emojis = Object.keys(data[category])
    return emojis.map((emoji) => [emoji, category, ...(data[category][emoji])])
  }).flat(1)
  return {
    categories,
    userLangCode,
    noteLangCode,
    flag: locale_flag,
    strings,
    notes,
    locales: locales
      .map((localeData) => localeData)
      .sort((a, b) => {
        if (a.locale_code < b.locale_code) return -1
        if (a.locale_code > b.locale_code) return 1
        return 0
      }),
  }
}
