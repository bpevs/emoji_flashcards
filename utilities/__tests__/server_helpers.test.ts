import { pageHandler } from '../server_helpers.ts'
import { assertExists } from 'std/assert/mod.ts'
import { it } from 'std/testing/bdd.ts'
import { Application, Context } from 'oak'
const app = new Application()

it('should render index page', async () => {
  const handler = pageHandler('index')
  const request = {
    url: new URL('https://flashcards.bpev.me/?card=el-GR&i=0'),
    headers: new Headers(),
    getBody: () => '',
  }
  // deno-lint-ignore no-explicit-any
  const context = new Context(app, request as any, {})
  await handler(context)
  assertExists(context.response.body)
})
