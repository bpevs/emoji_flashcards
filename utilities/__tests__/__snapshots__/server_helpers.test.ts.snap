export const snapshot = {};

snapshot[`should render index page 1`] = `
'
              <option value="af-ZA">
                🇿🇦 Afrikaans
              </option>
              <option value="ar-EG">
                🇪🇬 الأفريقانية
              </option>
              <option value="bg-BG">
                🇧🇬 Български
              </option>
              <option value="ca-ES">
                🇪🇸 Català
              </option>
              <option value="cs-CZ">
                🇨🇿 Čeština
              </option>
              <option value="da-DK">
                🇩🇰 Dansk
              </option>
              <option value="de-DE">
                🇩🇪 Deutsch
              </option>
              <option value="el-GR">
                🇬🇷 Ελληνικά
              </option>
              <option value="en-GB">
                🇬🇧 English (British)
              </option>
              <option value="en-US" selected="selected">
                🇺🇸 English (American)
              </option>
              <option value="es-ES">
                🇪🇸 Español (España)
              </option>
              <option value="es-MX">
                🇲🇽 Español (México)
              </option>
              <option value="et-EE">
                🇪🇪 Eesti
              </option>
              <option value="fa-IR">
                🇮🇷 فارسی
              </option>
              <option value="fi-FI">
                🇫🇮 Suomi
              </option>
              <option value="fr-FR">
                🇫🇷 Français
              </option>
              <option value="he-IL">
                🇮🇱 עברית
              </option>
              <option value="hi-IN">
                🇮🇳 हिंदी
              </option>
              <option value="hu-HU">
                🇭🇺 Magyar
              </option>
              <option value="id-ID">
                🇮🇩 Bahasa Indonesia
              </option>
              <option value="it-IT">
                🇮🇹 Italiano
              </option>
              <option value="ja-JP">
                🇯🇵 日本語
              </option>
              <option value="ko-KR">
                🇰🇷 한국어
              </option>
              <option value="lt-LT">
                🇱🇹 Lietuvių
              </option>
              <option value="lv-LV">
                🇱🇻 Latviešu
              </option>
              <option value="nb-NO">
                🇳🇴 Norsk
              </option>
              <option value="nl-NL">
                🇳🇱 Nederlands
              </option>
              <option value="pl-PL">
                🇵🇱 Polski
              </option>
              <option value="pt-BR">
                🇧🇷 Português (Brasil)
              </option>
              <option value="pt-PT">
                🇵🇹 Português (Portugal)
              </option>
              <option value="ro-RO">
                🇷🇴 Română
              </option>
              <option value="ru-RU">
                🇷🇺 Русский
              </option>
              <option value="sk-SK">
                🇸🇰 Slovinčina
              </option>
              <option value="sl-SI">
                🇸🇮 Slovenski
              </option>
              <option value="sv-SE">
                🇸🇪 Svenska
              </option>
              <option value="th-TH">
                🇹🇭 ไทย
              </option>
              <option value="tr-TR">
                🇹🇷 Turkiska
              </option>
              <option value="uk-UA">
                🇺🇦 Українська
              </option>
              <option value="ur-PK">
                🇵🇰 اردو
              </option>
              <option value="vi-VN">
                🇻🇳 Tiếng Việt
              </option>
              <option value="zh-CN">
                🇨🇳 中文（普通话）
              </option>
              <option value="zh-HK">
                🇭🇰 中文（粵語）
              </option>
              <option value="zh-TW">
                🇹🇼 中文（台灣）
              </option>
        '
`;

snapshot[`should render index page 2`] = `
'
              <option value="af-ZA">
                🇿🇦 Afrikaans
              </option>
              <option value="ar-EG">
                🇪🇬 Arabic
              </option>
              <option value="bg-BG">
                🇧🇬 Bulgarian
              </option>
              <option value="ca-ES">
                🇪🇸 Catalan
              </option>
              <option value="cs-CZ">
                🇨🇿 Czech
              </option>
              <option value="da-DK">
                🇩🇰 Danish
              </option>
              <option value="de-DE">
                🇩🇪 German
              </option>
              <option value="el-GR">
                🇬🇷 Greek
              </option>
              <option value="en-GB">
                🇬🇧 English (British)
              </option>
              <option value="en-US" selected="selected">
                🇺🇸 English (American)
              </option>
              <option value="es-ES">
                🇪🇸 Spanish (Spain)
              </option>
              <option value="es-MX">
                🇲🇽 Spanish (Mexico)
              </option>
              <option value="et-EE">
                🇪🇪 Estonian
              </option>
              <option value="fa-IR">
                🇮🇷 Persian
              </option>
              <option value="fi-FI">
                🇫🇮 Finnish
              </option>
              <option value="fr-FR">
                🇫🇷 French
              </option>
              <option value="he-IL">
                🇮🇱 Hebrew
              </option>
              <option value="hi-IN">
                🇮🇳 Hindi
              </option>
              <option value="hu-HU">
                🇭🇺 Hungarian
              </option>
              <option value="id-ID">
                🇮🇩 Indonesian
              </option>
              <option value="it-IT">
                🇮🇹 Italian
              </option>
              <option value="ja-JP">
                🇯🇵 Japanese
              </option>
              <option value="ko-KR">
                🇰🇷 Korean
              </option>
              <option value="lt-LT">
                🇱🇹 Lithuanian
              </option>
              <option value="lv-LV">
                🇱🇻 Latvian
              </option>
              <option value="nb-NO">
                🇳🇴 Norwegian
              </option>
              <option value="nl-NL">
                🇳🇱 Dutch
              </option>
              <option value="pl-PL">
                🇵🇱 Polish
              </option>
              <option value="pt-BR">
                🇧🇷 Portuguese (Brazil)
              </option>
              <option value="pt-PT">
                🇵🇹 Portuguese (Portugal)
              </option>
              <option value="ro-RO">
                🇷🇴 Romanian
              </option>
              <option value="ru-RU">
                🇷🇺 Russian
              </option>
              <option value="sk-SK">
                🇸🇰 Slovak
              </option>
              <option value="sl-SI">
                🇸🇮 Slovenian
              </option>
              <option value="sv-SE">
                🇸🇪 Swedish
              </option>
              <option value="th-TH">
                🇹🇭 Thai
              </option>
              <option value="tr-TR">
                🇹🇷 Turkish
              </option>
              <option value="uk-UA">
                🇺🇦 Ukranian
              </option>
              <option value="ur-PK">
                🇵🇰 Urdu
              </option>
              <option value="vi-VN">
                🇻🇳 Vietnamese
              </option>
              <option value="zh-CN">
                🇨🇳 Chinese (Mandarin)
              </option>
              <option value="zh-HK">
                🇭🇰 Chinese (Cantonese)
              </option>
              <option value="zh-TW">
                🇹🇼 Chinese (Taiwan)
              </option>
        '
`;

snapshot[`should render note selector in user language 1`] = `
'
              <option value="af-ZA">
                🇿🇦 Afrikaans
              </option>
              <option value="ar-EG">
                🇪🇬 Αραβικά
              </option>
              <option value="bg-BG">
                🇧🇬 Βουλγαρικά
              </option>
              <option value="ca-ES">
                🇪🇸 Καταλανικά
              </option>
              <option value="cs-CZ">
                🇨🇿 τσεχικά
              </option>
              <option value="da-DK">
                🇩🇰 Δανικά
              </option>
              <option value="de-DE">
                🇩🇪 Γερμανικά
              </option>
              <option value="el-GR">
                🇬🇷 Ελληνικά
              </option>
              <option value="en-GB">
                🇬🇧 Αγγλικά (βρετανικά)
              </option>
              <option value="en-US" selected="selected">
                🇺🇸 Αγγλικά (αμερικανικά)
              </option>
              <option value="es-ES">
                🇪🇸 Ισπανικά (Ισπανία)
              </option>
              <option value="es-MX">
                🇲🇽 Ισπανικά (Μεξικό)
              </option>
              <option value="et-EE">
                🇪🇪 Εσθονικά
              </option>
              <option value="fa-IR">
                🇮🇷 Περσικά
              </option>
              <option value="fi-FI">
                🇫🇮 Φινλανδικά
              </option>
              <option value="fr-FR">
                🇫🇷 Γαλλικά
              </option>
              <option value="he-IL">
                🇮🇱 Εβραϊκά
              </option>
              <option value="hi-IN">
                🇮🇳 Χίντι
              </option>
              <option value="hu-HU">
                🇭🇺 Ουγγρικά
              </option>
              <option value="id-ID">
                🇮🇩 Ινδονησιακά
              </option>
              <option value="it-IT">
                🇮🇹 Ιταλικά
              </option>
              <option value="ja-JP">
                🇯🇵 Ιαπωνικά
              </option>
              <option value="ko-KR">
                🇰🇷 Κορεάτικα
              </option>
              <option value="lt-LT">
                🇱🇹 Λιθουανικά
              </option>
              <option value="lv-LV">
                🇱🇻 Λετονία
              </option>
              <option value="nb-NO">
                🇳🇴 Νορβηγικά
              </option>
              <option value="nl-NL">
                🇳🇱 Ολλανδικά
              </option>
              <option value="pl-PL">
                🇵🇱 Πολωνικά
              </option>
              <option value="pt-BR">
                🇧🇷 Πορτογαλικά (Βραζιλία)
              </option>
              <option value="pt-PT">
                🇵🇹 Πορτογαλικά (Πορτογαλία)
              </option>
              <option value="ro-RO">
                🇷🇴 Ρουμανικά
              </option>
              <option value="ru-RU">
                🇷🇺 Ρωσικά
              </option>
              <option value="sk-SK">
                🇸🇰 Σλοβάκικα
              </option>
              <option value="sl-SI">
                🇸🇮 Σλοβένικα
              </option>
              <option value="sv-SE">
                🇸🇪 Σουηδικά
              </option>
              <option value="th-TH">
                🇹🇭 Ταϊλάνδη
              </option>
              <option value="tr-TR">
                🇹🇷 Τουρκικά
              </option>
              <option value="uk-UA">
                🇺🇦 ουκρανικά
              </option>
              <option value="ur-PK">
                🇵🇰 Ουρντού
              </option>
              <option value="vi-VN">
                🇻🇳 βιετναμέζικα
              </option>
              <option value="zh-CN">
                🇨🇳 Κινέζικα (μανδαρινικά)
              </option>
              <option value="zh-HK">
                🇭🇰 Κινέζικα (Καντονέζικα)
              </option>
              <option value="zh-TW">
                🇹🇼 Κινέζικα (Ταϊβάν)
              </option>
        '
`;
